File: python/seed_and_extend.py
"""
Seed-and-Extend Algorithm using K-mer Hashing

This algorithm is used for fast sequence alignment by:
1. Finding exact k-mer matches (seeds) using hash tables
2. Extending seeds to find longer alignments

This is the foundational approach used in BLAST, MAQ, and SOAP.
"""

from collections import defaultdict


class KmerIndex:
    """Index for storing k-mer positions in a reference sequence."""
    
    def __init__(self, sequence, k=11):
        """
        Build k-mer index from a reference sequence.
        
        Args:
            sequence: Reference sequence (string)
            k: K-mer length (default: 11)
        """
        self.sequence = sequence
        self.k = k
        self.index = defaultdict(list)
        self._build_index()
    
    def _build_index(self):
        """Build hash table index of k-mer positions."""
        for i in range(len(self.sequence) - self.k + 1):
            kmer = self.sequence[i:i + self.k]
            self.index[kmer].append(i)
    
    def find_seeds(self, query):
        """
        Find exact k-mer matches between query and reference.
        
        Args:
            query: Query sequence (string)
        
        Returns:
            list: List of (query_pos, ref_pos) tuples for each seed match
        """
        seeds = []
        for i in range(len(query) - self.k + 1):
            kmer = query[i:i + self.k]
            if kmer in self.index:
                for ref_pos in self.index[kmer]:
                    seeds.append((i, ref_pos))
        return seeds


def extend_seed(seq1, seq2, seed_pos1, seed_pos2, match_score=1, mismatch_penalty=-1):
    """
    Extend a seed match in both directions.
    
    Args:
        seq1: First sequence (string)
        seq2: Second sequence (string)
        seed_pos1: Starting position in seq1
        seed_pos2: Starting position in seq2
        match_score: Score for matching characters
        mismatch_penalty: Penalty for mismatching characters
    
    Returns:
        tuple: (start1, end1, start2, end2, score)
    """
    # Extend right
    score = 0
    max_score = 0
    max_right = 0
    
    i = 0
    while (seed_pos1 + i < len(seq1) and 
           seed_pos2 + i < len(seq2)):
        if seq1[seed_pos1 + i] == seq2[seed_pos2 + i]:
            score += match_score
        else:
            score += mismatch_penalty
        
        if score > max_score:
            max_score = score
            max_right = i + 1
        
        # Stop if score drops too much below max
        if score < max_score - 5:
            break
        
        i += 1
    
    # Extend left
    score = max_score
    max_left = 0
    
    i = 1
    while (seed_pos1 - i >= 0 and 
           seed_pos2 - i >= 0):
        if seq1[seed_pos1 - i] == seq2[seed_pos2 - i]:
            score += match_score
        else:
            score += mismatch_penalty
        
        if score > max_score:
            max_score = score
            max_left = i
        
        # Stop if score drops too much below max
        if score < max_score - 5:
            break
        
        i += 1
    
    start1 = seed_pos1 - max_left
    end1 = seed_pos1 + max_right
    start2 = seed_pos2 - max_left
    end2 = seed_pos2 + max_right
    
    return start1, end1, start2, end2, max_score


def seed_and_extend(reference, query, k=11):
    """
    Perform seed-and-extend alignment between query and reference.
    
    Args:
        reference: Reference sequence (string)
        query: Query sequence (string)
        k: K-mer length (default: 11)
    
    Returns:
        list: List of alignment hits with (query_start, query_end, ref_start, ref_end, score)
    """
    # Build k-mer index
    kmer_index = KmerIndex(reference, k)
    
    # Find seeds
    seeds = kmer_index.find_seeds(query)
    
    # Extend each seed
    alignments = []
    for query_pos, ref_pos in seeds:
        start1, end1, start2, end2, score = extend_seed(
            query, reference, query_pos, ref_pos
        )
        alignments.append({
            'query_start': start1,
            'query_end': end1,
            'ref_start': start2,
            'ref_end': end2,
            'score': score,
            'query_seq': query[start1:end1],
            'ref_seq': reference[start2:end2]
        })
    
    # Sort by score descending
    alignments.sort(key=lambda x: x['score'], reverse=True)
    
    return alignments


if __name__ == "__main__":
    # Example usage
    reference = "ACGTACGTACGTAAACCCGGGTTTACGTACGT"
    query = "ACGTAAACCCGGG"
    
    print("Seed-and-Extend Alignment")
    print(f"Reference: {reference}")
    print(f"Query: {query}")
    print(f"\nK-mer length: 5")
    
    alignments = seed_and_extend(reference, query, k=5)
    
    print(f"\nFound {len(alignments)} alignment(s):")
    for i, aln in enumerate(alignments[:5], 1):  # Show top 5
        print(f"\nAlignment {i}:")
        print(f"  Query position: {aln['query_start']}-{aln['query_end']}")
        print(f"  Reference position: {aln['ref_start']}-{aln['ref_end']}")
        print(f"  Score: {aln['score']}")
        print(f"  Query sequence: {aln['query_seq']}")
        print(f"  Ref sequence:   {aln['ref_seq']}")

File: cpp/seed_and_extend.cpp
/**
 * Seed-and-Extend Algorithm using K-mer Hashing
 * 
 * This algorithm is used for fast sequence alignment by:
 * 1. Finding exact k-mer matches (seeds) using hash tables
 * 2. Extending seeds to find longer alignments
 * 
 * This is the foundational approach used in BLAST, MAQ, and SOAP.
 */

#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

struct Seed {
    int query_pos;
    int ref_pos;
};

struct AlignmentHit {
    int query_start;
    int query_end;
    int ref_start;
    int ref_end;
    int score;
    string query_seq;
    string ref_seq;
};

/**
 * Build k-mer index from reference sequence.
 */
unordered_map<string, vector<int>> build_kmer_index(const string& sequence, int k) {
    unordered_map<string, vector<int>> index;
    
    for (size_t i = 0; i <= sequence.length() - k; i++) {
        string kmer = sequence.substr(i, k);
        index[kmer].push_back(i);
    }
    
    return index;
}

/**
 * Find exact k-mer matches between query and reference.
 */
vector<Seed> find_seeds(const string& query, 
                       const unordered_map<string, vector<int>>& index, 
                       int k) {
    vector<Seed> seeds;
    
    for (size_t i = 0; i <= query.length() - k; i++) {
        string kmer = query.substr(i, k);
        
        auto it = index.find(kmer);
        if (it != index.end()) {
            for (int ref_pos : it->second) {
                seeds.push_back({(int)i, ref_pos});
            }
        }
    }
    
    return seeds;
}

/**
 * Extend a seed match in both directions.
 */
AlignmentHit extend_seed(const string& seq1, const string& seq2,
                        int seed_pos1, int seed_pos2,
                        int match_score = 1, int mismatch_penalty = -1) {
    int len1 = seq1.length();
    int len2 = seq2.length();
    
    // Extend right
    int score = 0;
    int max_score = 0;
    int max_right = 0;
    
    int i = 0;
    while (seed_pos1 + i < len1 && seed_pos2 + i < len2) {
        if (seq1[seed_pos1 + i] == seq2[seed_pos2 + i]) {
            score += match_score;
        } else {
            score += mismatch_penalty;
        }
        
        if (score > max_score) {
            max_score = score;
            max_right = i + 1;
        }
        
        // Stop if score drops too much below max
        if (score < max_score - 5) {
            break;
        }
        
        i++;
    }
    
    // Extend left
    score = max_score;
    int max_left = 0;
    
    i = 1;
    while (seed_pos1 - i >= 0 && seed_pos2 - i >= 0) {
        if (seq1[seed_pos1 - i] == seq2[seed_pos2 - i]) {
            score += match_score;
        } else {
            score += mismatch_penalty;
        }
        
        if (score > max_score) {
            max_score = score;
            max_left = i;
        }
        
        // Stop if score drops too much below max
        if (score < max_score - 5) {
            break;
        }
        
        i++;
    }
    
    int start1 = seed_pos1 - max_left;
    int end1 = seed_pos1 + max_right;
    int start2 = seed_pos2 - max_left;
    int end2 = seed_pos2 + max_right;
    
    return {
        start1, end1, start2, end2, max_score,
        seq1.substr(start1, end1 - start1),
        seq2.substr(start2, end2 - start2)
    };
}

/**
 * Perform seed-and-extend alignment between query and reference.
 */
vector<AlignmentHit> seed_and_extend(const string& reference, 
                                    const string& query, 
                                    int k = 11) {
    // Build k-mer index
    auto index = build_kmer_index(reference, k);
    
    // Find seeds
    vector<Seed> seeds = find_seeds(query, index, k);
    
    // Extend each seed
    vector<AlignmentHit> alignments;
    for (const auto& seed : seeds) {
        AlignmentHit hit = extend_seed(query, reference, seed.query_pos, seed.ref_pos);
        alignments.push_back(hit);
    }
    
    // Sort by score descending
    sort(alignments.begin(), alignments.end(), 
         [](const AlignmentHit& a, const AlignmentHit& b) {
             return a.score > b.score;
         });
    
    return alignments;
}

int main() {
    // Example usage
    string reference = "ACGTACGTACGTAAACCCGGGTTTACGTACGT";
    string query = "ACGTAAACCCGGG";
    
    cout << "Seed-and-Extend Alignment" << endl;
    cout << "Reference: " << reference << endl;
    cout << "Query: " << query << endl;
    cout << "\nK-mer length: 5" << endl;
    
    vector<AlignmentHit> alignments = seed_and_extend(reference, query, 5);
    
    cout << "\nFound " << alignments.size() << " alignment(s):" << endl;
    
    // Show top 5
    for (size_t i = 0; i < min(size_t(5), alignments.size()); i++) {
        const auto& aln = alignments[i];
        cout << "\nAlignment " << (i + 1) << ":" << endl;
        cout << "  Query position: " << aln.query_start << "-" << aln.query_end << endl;
        cout << "  Reference position: " << aln.ref_start << "-" << aln.ref_end << endl;
        cout << "  Score: " << aln.score << endl;
        cout << "  Query sequence: " << aln.query_seq << endl;
        cout << "  Ref sequence:   " << aln.ref_seq << endl;
    }
    
    return 0;
}
